商品出品機能

donrow007 opened the pull request:
what
商品出品機能の実装（ActiveHash・Active Storage・JavaScriptの活用）
モデル単体テストの実装

why
ユーザーが、販売したい商品の登録をするため
データベースに保存する際、保存できないデータを入力していないか確認をするため。

ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/14362ad0de3879937207d5ac17fc3a86


価格が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/0683bbcdf57db85a622b5bfff598c843


必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/e683ae14922af288f78d58a3a06afc85


入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/a1bf65e9b2b83f94eb59fae276c21ed9


ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/e504f3f6b3d85ebf17385f711705210e


テスト結果の画像（bundle exec rspec spec/models/item_spec.rb）
https://gyazo.com/5d6fa6f77cf9d0394c303b0debefb3ef

donrow007 committed:
itemモデルの作成

rails g model itemにて作成。

donrow007 committed:
商品出品情報を保存するためのテーブル作成。

donrow007 committed:
ActiveHashの実装。

　データベースを作成しないカテゴリー、商品の状態、配送料の負担、発送元の地域、発送までの日数についてActiveHashを活用した。
　nameカラムを作りわすれたので% rails db:rollbackする。
　shipping_day_idへ修正（モデルは単数形の為。合わせて、READMEとER図も修正）

donrow007 committed:
mini_magick、image_processing、active_hash（Gem）の追加。

donrow007 committed:
itemのビューファイルについて

カラムに合わせたname属性・ActiveHashにて作成したクラス名・エラー発生時にメッセージが表示されるよう実装。

donrow007 committed:
itemモデルの実装とuserモデルについて修正。

アソシエーションを実装したり、itemモデルに関してはバリデーションをする。

donrow007 committed:
新規登録するとエラーが起こる

消去すべきものが残っていた為

donrow007 committed:
出品画面に遷移しない

ディレクトリ名とクラス名の相違があったため；Name Error

donrow007 committed:
画像を保存するため

Active Storageのテーブルで管理された画像ファイルのアソシエーション

donrow007 committed:
出品の情報が保存されない

入力ミスのため

donrow007 committed:
itemモデルテストコードの実装。

イメージ画像（テスト用）を実装。
active_storageのマイグレーションファイルをDBへ反映。

donrow007 committed:
単体テストコード前のエラー

余計なendがあったため。

donrow007 committed:
テストコードのエラーがたくさん

参考WEBより
可能性としてはテストに負荷がかかりすぎてread_timeout以内に処理が終わらないため。
　インスタンスを生成する処理に0.1秒待機させてread_timeoutのエラーを回避するための記述。

donrow007 committed:
Item 商品の出品登録 出品ができないとき １枚画像がないと出品できないエラー

validates :image, presence: trueをする。

donrow007 committed:
販売手数料・利益をJavaScriptで実装。

donrow007 committed:
プルリクエスト前の最終調整

yuri240 commented on app/models/item.rb:
```diff
@@ -0,0 +1,40 @@
+  has_one_attached :image
+
+  
+    validates :user_id, presence: true 
```
```ruby
validates :user_id, presence: true 
```
userに対するバリデーションは不要です。
19行目は削除しましょう！

理由：
`belongs_to :user`があることで自動的に「必ず存在しなければならない」というバリデーションが付与される為。

yuri240 commented on app/models/item.rb:
```diff
@@ -0,0 +1,40 @@
+    validates :prefecture_id, presence: true 
+    validates :shipping_day_id, presence: true 
+    # 300円以上かつ9,999,999円以下で、半角数字でないと入力不可
+    validates :price, numericality: { greater_than_or_equal_to: 300, less_than_or_equal_to: 9_999_999  }
```
priceカラムには、**整数のみ**保存が行われるようバリデーションを追加しましょう！

理由：
現状のコードですと、小数でもバリデーションが通ってしまうため。

yuri240 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,127 @@
+RSpec.describe Item, type: :model do
+  before do
+    @item = FactoryBot.create(:item)
+    sleep 0.1 
```
商品出品機能の単体テストにおいては、
FactoryBotで生成したデータを都度データベースへ保存する必要がございません。

したがって、`create`メソッドを使用するのではなく、
`build`メソッドを使用していただくようお願いいたします。

理由：
buildメソッドを使用することで、データベースへの負荷が減り、
5行目の`sleep 0.1`も不要となるためです。

yuri240 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,127 @@
+    it 'ユーザー登録している人でないと出品できない' do
+      @item.user_id = nil
+      @item.valid?
+      expect(@item.errors.full_messages).to include('User must exist', "User can't be blank")
```
```ruby
include('User must exist', "User can't be blank")
```
こちら1つのテストに対して1つの期待するエラーメッセージを設定する原則がございます。
適切なエラーメッセージのみ記述しましょう！

yuri240 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,127 @@
+    it '配送料の負担の情報が空欄だと出品できない' do
+      @item.shipping_cost_id = nil
+      @item.valid?
+      expect(@item.errors.full_messages).to include("Shipping cost can't be blank", 'Shipping cost is not a number')
```
```ruby
include("Category can't be blank", 'Category is not a number')
```
アクティブハッシュのテストに関しても同様に
1つのテストに対して1つの期待するエラーメッセージを設定する原則がございますので、
適切なエラーメッセージのみ記述しましょう！

yuri240 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,127 @@
+      @item.valid?
+      expect(@item.errors.full_messages).to include("Shipping day can't be blank", 'Shipping day is not a number')
+    end
+    it '価格の情報がないと出品できない' do
```
priceのテスト項目について、以下を追加で実行しましょう！

```
・価格に半角数字以外が含まれている場合は出品できない
```

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
コメントさせていただきましたので、ご確認お願いいたします。

donrow007 committed:
商品出品機能　修正1回目

・19行目は削除→削除しました。

・整数のみ保存が行われるようバリデーション→validates :price, numericality: { only_integer: true }設定しました。

・sleep 0.1→削除しました。

・include('User must exist', "User can't be blank")→ログインしないで出品すると、エラーがでるので削除しました。

・1つのテストに対して1つの期待するエラーメッセージ→各項目の「---」と空欄のバリデーションが重なるので、デフォルトの「---」を優先して、空欄のタグは削除しました。

・価格に半角数字以外が含まれている場合は出品できない→追加致しました。

よろしくお願い致します。

donrow007 commented:
商品出品機能　修正１回目
テスト結果の画像
https://gyazo.com/5a92135cda657ee0ef64cc5652760616

よろしくお願い致します。

yuri240 commented on app/models/category.rb:
```diff
@@ -0,0 +1,18 @@
+class Category < ActiveHash::Base
+  self.data = [
+    { id: 0, name: '--' },
```
```
 { id: 0, name: '--' },
```
見本アプリ同様に未選択は「` '---' `」としましょう！

yuri240 commented on app/models/item_condition.rb:
```diff
@@ -0,0 +1,14 @@
+class ItemCondition < ActiveHash::Base
+  self.data = [
+    { id: 0, name: '--' },
```
```
{ id: 0, name: '--' },
```
上記も「`'---'`」に修正をお願いいたします。

yuri240 commented on app/models/item_condition.rb:
```diff
@@ -0,0 +1,14 @@
+  self.data = [
+    { id: 0, name: '--' },
+    { id: 1, name: '新品、未使用' },
+    { id: 2, name: '未使用品に近い' },
```
```
    { id: 1, name: '新品、未使用' },
    { id: 2, name: '未使用品に近い' },
```
上記についても見本アプリ同様の記述に修正しましょう！
↓

```
    { id: 1, name: '新品・未使用' },
    { id: 2, name: '未使用に近い' },
```

yuri240 commented on app/models/shipping_cost.rb:
```diff
@@ -0,0 +1,10 @@
+  self.data = [
+    { id: 0, name: '--' },
+    { id: 1, name: '着払い（購入者負担）' },
+    { id: 2, name: '送料払い（出品者負担）' }
```
3行目の修正（「`'---'`」）に加えて、5行目は「`送料込み（出品者負担）`」に修正をお願いいたします。

yuri240 commented on app/models/shipping_day.rb:
```diff
@@ -0,0 +1,11 @@
+class ShippingDay < ActiveHash::Base
+  self.data = [
+    { id: 0, name: '--' },
```
上記も「`'---'`」に修正をお願いいたします。

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
キャプチャ画像と修正点確認させていただきました。

追加でコメントさせていただきましたので、ご確認お願いします。

あと一息かと思いますので、一緒に頑張りましょう！

donrow007 committed:
商品出品機能　修正2回目

修正致しました。
よろしくお願い致します。

div-reviewer reviewed (commented):
donrow007様、学習お疲れ様です！
お忙しい中商品出品の再レビュー依頼ありがとうございます。プルリクエスト確認させていただきました。

Activehashのモデルに記載しているデータが綺麗に記載できましたね。この調子で最後まで頑張りましょう。
修正点をコメントさせて頂きましたので、ご確認をお願い致します。

```ruby
【次回プルリクエスト提出する際の注意点】
①修正が終わりましたら、`コミットとプッシュ`をして、`プルリクエストに修正内容が反映されているか確認`しましょう。
②既に頂いている挙動確認の画像や動画が修正範囲だった場合、最新の画像や動画の提出をお願い致します。
③googleフォームの送信をお願いいたします。
```


ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞよろしくおねがいします。

donrow007 committed:
商品出品機能　修正３回目

修正致しました。
よろしくお願い致します。

donrow007 commented:
テストコードの動画
https://gyazo.com/cca41e25a145fc3e80e2e3defcfedd4e
よろしくお願い致します。

sh10n7 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,103 @@
+      expect(@item.errors.full_messages).to include("Price can't be blank")
+    end
+    it '価格に半角数字以外が含まれている場合は出品できない' do
+      @item.price = nil
```
```
@item.price = nil
```
テストのタイトル(```価格に半角数字以外が含まれている場合は出品できない'```)と、テスト内容に相違がございます。

上記の記述ですと、価格が空(nil)だった場合のテストとなりますので、@item.priceに全角文字などを代入するように記述を変更しましょう。

sh10n7 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,103 @@
+      @item.valid?
+      expect(@item.errors.full_messages).to include('Shipping day must be other than 0')
+    end
+    
```
下記項目のテストコードも作成しましょう。
```
■ActiveHash５項目
・発送元の地域に「---」が選択されている場合は出品できない
```

理由：
異常系テストコードの要件を満たすためです

donrow007 replied:
お忙しいところ、恐れいります。
勉強不足ですが、上記の設定だと問題があるのでしょうか？ヒントを教えて頂けますでしょうか？

sh10n7 replied:
[こちら](https://github.com/donrow007/furima-38283/pull/15/files#r1032190741)の修正が確認できませんでした。
再度ご確認の上、テストの追加をお願いいたします。

sh10n7 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。

追加で修正いただきたい箇所をコメントしております。ご確認いただけますと幸いです。
​
修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！

その際、**修正後のテスト実行結果の画像**も添付いただきますようお願いいたします。

ご不明点や疑問点ございましたら、チャット・コネクトメンターにご相談ください！
どうぞ、よろしくお願いいたします。

donrow007 committed:
商品出品機能　修正４回目

修正しました。
よろしくお願い致します。

donrow007 commented:
テストコード画像
https://gyazo.com/31c78e1875ec0bbbfd6b47c3ad732f9b
よろしくお願い致します。

sh10n7 reviewed (commented):
レビュー依頼ありがとうございます。
一部修正が確認できない箇所がございましたので、再度コメントをしております。
​
修正頂いた後は、再度レビュー依頼のご提出をお願い致します！
※併せて、**修正後のテスト実行結果の画像**も添付くださいますようお願いいたします。

ご対応のほどよろしくお願いいたします。

donrow007 committed:
商品出品機能　修正5回目

donrow007 commented:
テストコード
https://gyazo.com/15bd2436de4e4671b9d4073ba64286e5

発送の地域を追加しました。

よろしくお願い致します。

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
お送り頂いた内容を確認致しました。

今回【LGTM】とさせて頂きます。
実装お疲れ様でした！

```
※現在のブランチはマージ後、ローカルのmasterブランチに忘れずにプルしましょう！
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
```

引き続き実装頑張っていきましょう！
応援しております！

donrow007 commented:
　What
　ログイン有無に関わらず、トップページで出品登録した商品が表示されるように実装した。
①ログイン中
https://gyazo.com/eca9e353a57a06e81ef095578e3ee4df

②ログアウトした
https://gyazo.com/edf80f361c2b4967ad39742045d8ccc1

　商品のデータがない場合は、ダミー商品が表示されている動画
①データがある場合
https://gyazo.com/edc725d248fed32a687b4327f6774a0f

②データがない場合
https://gyazo.com/d1813d770a3a6b7864a7637b42d47ccc

　Why
トップページで出品した商品一覧を見られるようにするため。
 欲しい商品を見て新規登録を促すため。

よろしくお願い致します。

yuri240 commented:
レビュー依頼ありがとうございます。

こちらのプルリクエストは商品出品機能のものとなりますので、
新たに商品一覧表示機能のブランチを作成いただき、
そちらのプルリクエストをご提出いただけますでしょうか？

もしご不明点などございましたら
テックコネクトにてお気軽にお尋ねください。
